src/executables/sysrepocfg.c

```diff
@@ -0,0 +1,775 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <getopt.h>
+#include <dirent.h>
+#include <assert.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <fcntl.h>
+#include <ctype.h>
+#include <libyang/libyang.h>
+
+#include "sr_common.h"
+#include "client_library.h"
+
+#define EXPECTED_MAX_INPUT_FILE_SIZE  4096
+#define DEFAULT_TEXT_EDITOR           "vi"
+
+/**
+ * @brief Operation to be performed.
+ */
+typedef enum srcfg_operation_e {
+    SRCFG_OP_EDIT,   /**< Edit current configuration */
+    SRCFG_OP_IMPORT, /**< Import configuration from file or stdin */
+    SRCFG_OP_EXPORT  /**< Export configuration to file or stdout */
+} srcfg_operation_t;
+
+/**
+ * @brief Datastore to be operated on.
+ */
+typedef enum srcfg_datastore_e {
+    SRCFG_STORE_RUNNING,   /**< Work with the running datastore */
+    SRCFG_STORE_STARTUP    /**< Work with the startup datastore */
+} srcfg_datastore_t;
+
+static char *srcfg_schema_search_dir = SR_SCHEMA_SEARCH_DIR;
+static char *srcfg_data_search_dir = SR_DATA_SEARCH_DIR;
+static bool srcfg_custom_repository = false;
+static const char * const srcfg_default_text_editor = DEFAULT_TEXT_EDITOR;
+
+/**
+ * @brief Logging callback called from libyang for each log entry.
+ */
+static void
+srcfg_ly_log_cb(LY_LOG_LEVEL level, const char *msg, const char *path)
+{
+    return;
+}
+
+/**
+ * @brief Reads complete content of a file referenced by the descriptor 'fd' into the memory.
+ * Caller is responsible for deallocation of the memory block returned through the output argument 'out'.
+ * Returns SR_ERR_OK in case of success, SR_ERR_INTERNAL otherwise.
+ */
+static int
+srcfg_read_file_content(int fd, char **out)
+{
+    size_t size = EXPECTED_MAX_INPUT_FILE_SIZE;
+    unsigned cur = 0;
+    ssize_t n = 0;
+    char *buffer = malloc(size);
+    if (NULL == buffer) {
+        fprintf(stderr, "Error: Memory allocation failed.\n");
+        return SR_ERR_INTERNAL;
+    }
+
+    do {
+        if (size == cur + 1) {
+            size <<= 1;
+            char *new_buffer = realloc(buffer, size);
+            if (NULL == new_buffer) {
+                free(buffer);
+                fprintf(stderr, "Error: Memory allocation failed.\n");
+                return SR_ERR_INTERNAL;
+            }
+            buffer = new_buffer;
+        }
+        n = read(fd, buffer + cur, size - cur - 1);
+        if (0 > n) {
+            free(buffer);
+            fprintf(stderr, "Error: Read operation failed: %s.\n", strerror(errno));
+            return SR_ERR_INTERNAL;
+        }
+        cur += n;
+    } while (0 < n);
+
+    buffer[cur] = '\0';
+    *out = buffer;
+    return SR_ERR_OK;
+}
+/**
+ * @brief Initializes libyang ctx with all schemas installed for specified module in sysrepo.
+ */
+static int
+srcfg_ly_init(struct ly_ctx **ly_ctx, const char *module_name)
```

Comment:
Please add NULL checks for all arguments in all functions. It is a good practice to have them in each function - it can be very helpful by debugging. Have a look at macros in common/sr_helpers.h